### PR TITLE
=multiplex #139 more properly handle logLevel in multiplex handler

### PR DESCRIPTION
**Pull request by ktoso**
### Motivation:

Log level handling is a bit too naive in the multiplex handler when underlying loggers have varying levels.

### Modifications:

- effective log level has to be the lowest of all passed in handlers / same on level modification
- when dispatching, the multiplex handler has to filter again based on the _specific_ handler.

### Result:

- resolves https://github.com/apple/swift-log/issues/139

### CODE REVIEW

File: Tests/LoggingTests/LoggingTest.swift [hunk 1]
```diff
@@ -74,6 +74,73 @@ class LoggingTest: XCTestCase {
...
+                return [:]
+            }()
+        )
+
+        logging1.history.assertExist(level: .info, message: "info")
+        logging2.history.assertExist(level: .info, message: "info")
+
+        XCTAssertEqual(messageMaterializations, 1)
+        XCTAssertEqual(metadataMaterializations, 1)
+    }
```
Comment (ktoso) [inline Tests/LoggingTests/LoggingTest.swift:142]: this is kind of enforced "by construction" but wanted to have the test in anyway

File: Tests/LoggingTests/LoggingTest.swift [hunk 2]
```diff
@@ -74,6 +74,73 @@ class LoggingTest: XCTestCase {
...
+        logging1.history.assertNotExist(level: .trace, message: "trace")
+        logging1.history.assertNotExist(level: .debug, message: "debug")
+        logging1.history.assertExist(level: .info, message: "info")
+        logging1.history.assertExist(level: .warning, message: "warning")
+
+        logging2.history.assertNotExist(level: .trace, message: "trace")
+        logging2.history.assertExist(level: .debug, message: "debug")
+        logging2.history.assertExist(level: .info, message: "info")
+        logging2.history.assertExist(level: .warning, message: "warning")
+    }
```
Comment (ktoso) [inline Tests/LoggingTests/LoggingTest.swift:106]: this test explains the feature

File: Sources/Logging/Logging.swift [hunk 3]
```diff
@@ -472,24 +473,30 @@ public struct MultiplexLogHandler: LogHandler {
...
     public var logLevel: Logger.Level {
         get {
-            return self.handlers[0].logLevel
+            return self.effectiveLogLevel
         }
         set {
             self.mutatingForEachHandler {
                 $0.logLevel = newValue
             }
+            self.effectiveLogLevel = MultiplexLogHandler.effectiveLogLevel(self.handlers.map { $0.logLevel })
```
Comment (tomerd) [inline Sources/Logging/Logging.swift:487]: since we are setting all the handlers in the move loop, computing effectiveLogLevel seems redundant

File: Sources/Logging/Logging.swift [hunk 4]
```diff
@@ -472,24 +473,30 @@ public struct MultiplexLogHandler: LogHandler {
     public init(_ handlers: [LogHandler]) {
         assert(!handlers.isEmpty)
         self.handlers = handlers
+        self.effectiveLogLevel = MultiplexLogHandler.effectiveLogLevel(handlers.map { $0.logLevel })
```
Comment (tomerd) [inline Sources/Logging/Logging.swift:476]: +1

File: Sources/Logging/Logging.swift [hunk 5]
```diff
@@ -472,24 +473,30 @@ public struct MultiplexLogHandler: LogHandler {
...
         }
         set {
             self.mutatingForEachHandler {
                 $0.logLevel = newValue
             }
+            self.effectiveLogLevel = MultiplexLogHandler.effectiveLogLevel(self.handlers.map { $0.logLevel })
         }
     }
 
+    static func effectiveLogLevel(_ levels: [Logger.Level]) -> Logger.Level {
```
Comment (tomerd) [inline Sources/Logging/Logging.swift:491]: private?

File: Sources/Logging/Logging.swift [hunk 6]
```diff
@@ -472,24 +473,30 @@ public struct MultiplexLogHandler: LogHandler {
...
+        return levels.min() ?? .trace
+    }
+
     public func log(level: Logger.Level,
                     message: Logger.Message,
                     metadata: Logger.Metadata?,
                     file: String, function: String, line: UInt) {
-        self.handlers.forEach { handler in
+        for handler in self.handlers where handler.logLevel <= level {
```
Comment (tomerd) [inline Sources/Logging/Logging.swift:499]: wouldn't delegating to the underlying `handler.log` already filter by level? what am I missing?

File: Sources/Logging/Logging.swift (hunk 5 above)
Comment (ktoso) [reply Sources/Logging/Logging.swift:491]: ok will do

File: Sources/Logging/Logging.swift (hunk 3 above)
Comment (ktoso) [reply Sources/Logging/Logging.swift:487]: that was "problem", example:

```
logger1 at .info
logger2 at .debug

multiplex.debug() // would never be emitted
```

because:

`multiplex.loglevel` was `return self.handlers[0].logLevel` arbitrary using the "first" as the handlers log level.

How logger decides if it should pass value to handler is: 

```
if handler.logLevel <= level {
```

so even though the `logger2` _does_ want to log things on debug level, it would not, because `logger1` is set to info; That's a bit weird.

----

Solutions:

- (this PR) on construction find "lowest loglevel this multiplex handler should accept, because one of its parts wants to get it"
- OR: make it like a normal logger/handler that it defaults to `.info` and if we want to log at other levels we need to change that on the multiplex handler?
- OR: (how it works today) the "first logger" is special and decides at which log level we're logging.
  - note that `multiplex.logLevel = .trace` then is a bit weird -- we've just changed _all_ loggers to .trace even though before we were saying the "first one only matters"; so the current impl is confusing since without setting a level: handlers log on VARIOUS levels; with setting a level, ALL handlers log at the same level.
  - rather what we'd want to do IMO when setting `multiplex.logLevel = .trace` is that "this handler will now take trace logs; and if the log handlers it contains WANT TO (were passed in as such) to log at .trace they now have the chance to do so" (this PR) -- so we retain the log level configurations that were passed in...
  - perhaps we'd then need a way to _really set_ the underlying levels if we wanted to? not sure.

Comment (ktoso) [reply Sources/Logging/Logging.swift:487]: Oh wait, sorry didn't have my coffee yet! You mean that the looping here is redundant, thanks yes that's true.

☕ ☕ ☕

File: Sources/Logging/Logging.swift (hunk 6 above)
Comment (ktoso) [reply Sources/Logging/Logging.swift:499]: See the analysis here: https://github.com/apple/swift-log/pull/140/files#r447320481

Comment (ktoso) [reply Sources/Logging/Logging.swift:499]: > the underlying handler.log already filter by level? what am I missing?

No, a handler is not normally performing the log level check -- only the Logger.log() API does that:

```
    @inlinable
    public func log(level: Logger.Level,
                    _ message: @autoclosure () -> Logger.Message,
                    metadata: @autoclosure () -> Logger.Metadata? = nil,
                    file: String = #file, function: String = #function, line: UInt = #line) {
        if self.logLevel <= level {
            self.handler.log(level: level,
                             message: message(),
                             metadata: metadata(),
                             file: file, function: function, line: line)
        }
    }
```

and handlers are naive and "if i am called, it seems I should log." This is why we have to make the `multiplex.logLevel` smarter as we do in this PR:

- so the change to logLevel in this PR causes the `Logger.log()` to dispatch to this multiplex handler when _at least one of the log handlers would like to log this message_
- but since now we don't know which logger that was... we have to check.

There's an open question how `multiplex.logLevel = .info` relates to this all, since then all levels are made equal and all will log. But what about if we only do:

```
logger1 at debug
logger2 at info

multiplex.debug() // we only want it to be logged at debug at logger1, not at logger2 as well
```

Comment (ktoso) [reply Sources/Logging/Logging.swift:499]: For context, `StreamLogHandler`:

```swift
    public func log(level: Logger.Level,
                    message: Logger.Message,
                    metadata: Logger.Metadata?,
                    file: String, function: String, line: UInt) {
        let prettyMetadata = metadata?.isEmpty ?? true
            ? self.prettyMetadata
            : self.prettify(self.metadata.merging(metadata!, uniquingKeysWith: { _, new in new }))
```

Comment (tomerd) [reply Sources/Logging/Logging.swift:499]: ah you are correct, I forgot the filter was at the Logger level not backend one. this is 100%

**Comment by ktoso** [comment]
I now realized the current behavior is documented:

```
/// The first `LogHandler` passed to the initialisation function of `MultiplexLogHandler` control the `logLevel` as
/// well as the `metadata` for this `LogHandler`. Any subsequent `LogHandler`s used to initialise a
/// `MultiplexLogHandler` are merely to emit the log message to another place.
```

but it's admittably a bit weird... not sure what to do here tbh.

The current behavior has some upsides (detailed control) but it is somewhat unexpected as witnessed by #139 as well as my going down this rabbit hole.

Do we have opinions what do do here @glbrntt @Lukasa (optionally, only if you have time and want to chime in @weissi) @tomerd ?

**Comment by tomerd** [comment]
> Do we have opinions what do do here @glbrntt @Lukasa (optionally, only if you have time and want to chime in @weissi) @tomerd ?

for the log level, I think the semantics proposed by this PR is an improvement, but it does make the metadata story more murky since its still pinned to handler[0]

**Comment by Lukasa** [comment]
Yeah, I think this behavioural change needs to happen. The logical semantics of MultiplexLogHandler, in my mind, would be that its effective log level is the lowest of all log handlers it wraps, _and_ that it only dispatches the messages to log handlers whose actual log level is lower than that of the message. This makes it maximally useful.

**Comment by ktoso** [comment]
>  The logical semantics of MultiplexLogHandler, in my mind, would be that its effective log level is the lowest of all log handlers it wraps, and that it only dispatches the messages to log handlers whose actual log level is lower than that of the message.

Cool, very much agreed, that's what the current PR brings it to 👍 

I'll sanity check the metadata handling again here soon; I guess the semantics could be:

- do not just expose any of the underlying ones
- when set, set onto all underlying ones

or something like that; rather than the [0] based one

**Comment by glbrntt** [comment]
I agree with what's been said above: the current behaviour is weird; this change introduces the behaviour I would have expected.

File: Sources/Logging/Logging.swift [hunk 7]
```diff
@@ -458,40 +458,74 @@ extension Logger {
...
+/// Example:
+/// If log handler `A` is logging at `.debug` level, and log handler `B` is logging at `.info` level, the constructed
+/// `MultiplexLogHandler([A, B])`'s effective log level will be set to `.debug`, meaning that debug messages will be
+/// handled by this handler, while only logged by the underlying `A` log handler (since `B`'s log level is `.info`
+/// and thus it would not actually log that log message).
+///
+/// If the log level is _set_ on a `Logger` backed by an `MultiplexLogHandler` the log level will apply to *all*
+/// underlying log handlers. Allowing a logger to still select at what level it wants to log, regardless if the underlying
+/// handler is a multiplex or a normal one. If for some reason one might want to not allow changing a log level of a specific
+/// handled passed into the multiplex log handler, this is possible by wrapping it in a handler which ignores any log level changes.
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:476]: docs and examples on how the log level works

File: Sources/Logging/Logging.swift [hunk 8]
```diff
@@ -458,40 +458,74 @@ extension Logger {
...
+///
+/// Example:
+/// If the multiplex log handler was initiated with two handlers like this: `MultiplexLogHandler([handler1, handler2])`.
+/// The handlers each have some already set metadata: `handler1` has metadata values for keys `one` and `all`, and `handler2`
+/// has values for keys `two` and `all`.
+///
+/// A query through the multiplex log handler the key `one` naturally returns `handler1`'s value, and a query for `two`
+/// naturally returns `handler2`'s value. Querying for the key `all` will return `handler1`'s value, as that handler was indicated
+/// "more important" than the second handler. The same rule applies when querying for the `metadata` property of the
+/// multiplex log handler - it constructs `Metadata` uniquing values
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:503]: docs and examples how metadata works

Comment (ktoso) [reply Sources/Logging/Logging.swift:503]: I actually had a "smart and configurable" `EffectiveMetadataMode` based impl but I stepped back from it for a few hours to re-think and it was getting too complex IMO for what it is -- realistically this is not a "big problem that needs fancy solving" -- normally users just care about setting metadata on things and maybe some handlers have something specific to them, the conflicting cases should be rare and this simple rule is better than slamming more complexity into it.

// I even had an impl which would "collect all values if the same key is present in a few handlers into a logger metadata value Array etc... Way too fancy for what we're looking at I then realized.

File: Sources/Logging/Logging.swift [hunk 9]
```diff
@@ -458,40 +458,74 @@ extension Logger {
...
+    private var effectiveLogLevel: Logger.Level
 
     /// Create a `MultiplexLogHandler`.
     ///
     /// - parameters:
     ///    - handlers: An array of `LogHandler`s, each of which will receive the log messages sent to this `Logger`.
     ///                The array must not be empty.
     public init(_ handlers: [LogHandler]) {
-        assert(!handlers.isEmpty)
+        assert(!handlers.isEmpty, "MultiplexLogHandler.handlers MUST NOT be empty")
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:514]: explicit messages always better IMO

File: Sources/Logging/Logging.swift [hunk 10]
```diff
@@ -458,40 +458,74 @@ extension Logger {
...
         self.handlers = handlers
-        self.effectiveLogLevel = MultiplexLogHandler.effectiveLogLevel(handlers.map { $0.logLevel })
+        self.effectiveLogLevel = handlers.map({ $0.logLevel }).min() ?? .trace
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:516]: the ?? will never trigger but it's semantically "aha makes sense, we take the lowest" so I prefer that to writing `!` here

File: Sources/Logging/Logging.swift [hunk 11]
```diff
@@ -503,16 +537,27 @@ public struct MultiplexLogHandler: LogHandler {
...
     public var metadata: Logger.Metadata {
         get {
-            return self.handlers[0].metadata
+            var effectiveMetadata: Logger.Metadata = [:]
+            // as a rough estimate we assume that the underlying handlers have a similar metadata count,
+            // and we use the first one's current count to estimate how big of a dictionary we need to allocate:
+            effectiveMetadata.reserveCapacity(self.handlers.first!.metadata.count) // !-safe, we always have at least one handler
+            return self.handlers.reduce(into: effectiveMetadata) { effectiveMetadata, handler in
+                effectiveMetadata.merge(handler.metadata, uniquingKeysWith: { l, _ in l })
+            }
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:546]: We pick the "left" over any following ones "first handlers in the list are more important"

File: Sources/Logging/Logging.swift [hunk 12]
```diff
@@ -503,16 +537,27 @@ public struct MultiplexLogHandler: LogHandler {
...
-    public subscript(metadataKey metadataKey: String) -> Logger.Metadata.Value? {
+    public subscript(metadataKey metadataKey: Logger.Metadata.Key) -> Logger.Metadata.Value? {
         get {
-            return self.handlers[0].metadata[metadataKey]
+            for handler in handlers {
+                if let value = handler[metadataKey: metadataKey] {
+                    return value
+                }
+            }
+            return nil
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:560]: Again, implements the "first is most important"

**Comment by ktoso** [comment]
I poked at the metadata quite a bit over the weekend and think I've arrived at what is a good tradeoff here.

The metadata handling change is: cdb48692e8bd7aab029651336733da70fde17e6d

And please see the comments in-line. Short version though: we should not magically rely on the _first_ handler, but we should _prefer_ more-first passed in handler metadata values over latter ones. This is simple to explain and implement without some crazy things (I had some advanced implementations with merge strategies and whatnot, but they ended up feeling wrong in their own ways, so I believe a simple rule like this (see added docs) is a good tradeoff, esp since we should not expect to be hitting this in reality much -- it is the underlying handlers doing their job, and barely anyone should be reading from the logger metadata to "just have a look at it" after all).

File: Sources/Logging/Logging.swift [hunk 13]
```diff
@@ -458,54 +458,106 @@ extension Logger {
...
+/// Example:
+/// If log handler `A` is logging at `.debug` level, and log handler `B` is logging at `.info` level, the constructed
+/// `MultiplexLogHandler([A, B])`'s effective log level will be set to `.debug`, meaning that debug messages will be
+/// handled by this handler, while only logged by the underlying `A` log handler (since `B`'s log level is `.info`
+/// and thus it would not actually log that log message).
+///
+/// If the log level is _set_ on a `Logger` backed by an `MultiplexLogHandler` the log level will apply to *all*
+/// underlying log handlers. Allowing a logger to still select at what level it wants to log, regardless if the underlying
+/// handler is a multiplex or a normal one. If for some reason one might want to not allow changing a log level of a specific
+/// handled passed into the multiplex log handler, this is possible by wrapping it in a handler which ignores any log level changes.
```
Comment (glbrntt) [inline Sources/Logging/Logging.swift:476]:
```suggestion
/// handler passed into the multiplex log handler, this is possible by wrapping it in a handler which ignores any log level changes.
```

File: Sources/Logging/Logging.swift [hunk 14]
```diff
@@ -458,54 +458,106 @@ extension Logger {
...
+///
+/// ### Effective Logger.Metadata
+///
+/// Since a `MultiplexLogHandler` is a combination of multiple log handlers, the handling of metadata can be non-obvious
+/// at first when the underlying log handlers have some metadata of their own set, before they are used to initialize the handler.
```
Comment (glbrntt) [inline Sources/Logging/Logging.swift:481]: I found this paragraph hard to parse:

```suggestion
/// Since a `MultiplexLogHandler` is a combination of multiple log handlers, the handling of metadata can be non-obvious.
/// For example, the underlying log handlers may have metadata of their own set before they are used to initialize the multiplex log handler.
```

File: Sources/Logging/Logging.swift [hunk 15]
```diff
@@ -458,54 +458,106 @@ extension Logger {
...
+/// any log events they might be interested in.
+///
+/// Example:
+/// If log handler `A` is logging at `.debug` level, and log handler `B` is logging at `.info` level, the constructed
+/// `MultiplexLogHandler([A, B])`'s effective log level will be set to `.debug`, meaning that debug messages will be
+/// handled by this handler, while only logged by the underlying `A` log handler (since `B`'s log level is `.info`
+/// and thus it would not actually log that log message).
+///
+/// If the log level is _set_ on a `Logger` backed by an `MultiplexLogHandler` the log level will apply to *all*
+/// underlying log handlers. Allowing a logger to still select at what level it wants to log, regardless if the underlying
```
Comment (Lukasa) [inline Sources/Logging/Logging.swift:474]: Nit, sentence fragment. Consider revising to:

```suggestion
/// underlying log handlers, allowing a logger to still select at what level it wants to log regardless of if the underlying
```

File: Sources/Logging/Logging.swift [hunk 16]
```diff
@@ -458,54 +458,106 @@ extension Logger {
...
+///
+/// Example:
+/// If the multiplex log handler was initiated with two handlers like this: `MultiplexLogHandler([handler1, handler2])`.
+/// The handlers each have some already set metadata: `handler1` has metadata values for keys `one` and `all`, and `handler2`
+/// has values for keys `two` and `all`.
+///
+/// A query through the multiplex log handler the key `one` naturally returns `handler1`'s value, and a query for `two`
+/// naturally returns `handler2`'s value. Querying for the key `all` will return `handler1`'s value, as that handler was indicated
+/// "more important" than the second handler. The same rule applies when querying for the `metadata` property of the
+/// multiplex log handler - it constructs `Metadata` uniquing values
```
Comment (glbrntt) [inline Sources/Logging/Logging.swift:503]:
```suggestion
/// multiplex log handler - it constructs `Metadata` uniquing values.
```

File: Sources/Logging/Logging.swift (hunk 14 above)
Comment (ktoso) [reply Sources/Logging/Logging.swift:481]: Thanks, I always over do it with long sentences 😅

**Comment by ktoso** [comment]
Thanks for the review and cleanups folks 👍
